src/pyodide-py/pyodide/console.py

```diff
@@ -0,0 +1,127 @@
+    ):
+        self._flush_callback = flush_callback
+        self.name = name
+
+    def writable(self):
+        return True
+
+    def seekable(self):
+        return False
+
+    def isatty(self):
+        return True
+
+    def write(self, data):
+        self._flush_callback(data.tobytes().decode())
+        return len(data)
+
+
+class InteractiveConsole(code.InteractiveConsole):
+    """Interactive Pyodide console
+
+    Base implementation for an interactive console that manages
+    stdout/stderr. Their value will be restored at destruction or
+    at `restore_stdstreams` call.
+
+    `self.stdout_callback` and `self.stderr_callback` can be overloaded.
+
+    Parameters
+    ----------
+    stdout_callback
+        Function to call at each `sys.stdout` flush.
+    stderr_callback
+        Function to call at each `sys.stderr` flush.
+    """
+
+    def __init__(
+        self,
+        locals: Optional[dict] = None,
+        stdout_callback: Optional[Callable[[str], None]] = None,
+        stderr_callback: Optional[Callable[[str], None]] = None,
+    ):
+        super().__init__(locals)
+        self.stdout_callback = stdout_callback
+        self.stderr_callback = stderr_callback
+        self.redirect_stdstreams()
+
+    def redirect_stdstreams(self):
+        """ Toggle stdout/stderr redirections. """
+
+        # we use meta callbacks to allow self.std{out,err}_callback
+        # overloading.
+        # we check callback against None at each call since it can be
+        # changed dynamically.
+
+        def meta_stdout_callback(*args):
+            if self.stdout_callback is not None:
+                return self.stdout_callback(*args)
+
+        def meta_stderr_callback(*args):
+            if self.stderr_callback is not None:
+                return self.stderr_callback(*args)
+
+        # for later restore
+        self._old_stdout = sys.stdout
+        self._old_stderr = sys.stderr
+
+        # it would be more robust to use sys.stdout.name and sys.stderr.name
+        # but testing system oveload them. Anyway it should be pretty stable
+        # upstream.
+        sys.stdout = _StdStream(meta_stdout_callback, name="<stdout>")
+        sys.stderr = _StdStream(meta_stderr_callback, name="<stderr>")
+
+    def restore_stdstreams(self):
+        """Restore stdout/stderr to the value it was before
+        the creation of the object."""
+        sys.stdout = self._old_stdout
+        sys.stderr = self._old_stderr
+
+    def __del__(self):
+        self.restore_stdstreams()
+
+    def banner(self):
+        """ A banner similar to the one printed by the real Python interpreter. """
+        # copyied from https://github.com/python/cpython/blob/799f8489d418b7f9207d333eac38214931bd7dcc/Lib/code.py#L214
+        cprt = 'Type "help", "copyright", "credits" or "license" for more information.'
+        return f"Python {sys.version} on {sys.platform}\n{cprt}"
```

Comment:
This would show "Python 3.8 on emscripten". Technically it's true but it might be confusing to users. I'm not sure what else we could put there, WebAssembly VM, pyodide, browser? Also keeping the snake emoji we had before would be a nice touch.

Reply:
It shows
```
Python 3.8.2 (default, Jan  4 2021, 12:17:49)
[Clang 6.0.1 (/b/s/w/ir/cache/git/chromium.googlesource.com-external-github.com on emscripten
Type "help", "copyright", "credits" or "license" for more information.
>>> 
```
which is not nice, you are right... 

I see something like

```
Python 3.8.2 (default, Jan  4 2021, 12:17:49) on WebAssembly VM
Type "help", "copyright", "credits" or "license" for more information.
```

for the default banner and

```
Python 3.8.2 (default, Jan  4 2021, 12:17:49) on WebAssembly VM
Welcome to the Pyodide terminal emulator 🐍
Type "help", "copyright", "credits" or "license" for more information.
```

for the one customized in `console.html`. WDYT?

Reply:
Or maybe:

```
Welcome to the Pyodide terminal emulator 🐍
Python 3.8.2 (default, Jan  4 2021, 12:17:49) on WebAssembly VM
Type "help", "copyright", "credits" or "license" for more information.
```

Reply:
I personally prefer the last version.

Reply:
Me too, this is the one I commited.